{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick start for binary classification using _ImaGene_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Tutorial to perform a binary classification using _ImaGene_. The aim is to classify whether a gene is under positive selection or neutrally evolving.\n",
    "\n",
    "The first step is to generate training and testing datasets. This is accomplished using _msms_. An example on how to do this is given in the script below. This script will split the simulations into different batches to later perform training with a \"simulation-on-the-fly\" approach."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bash Reproduce/generate_data.sh Binary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the necessary modules in _python_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import gzip\n",
    "import _pickle as pickle\n",
    "\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "\n",
    "import skimage.transform\n",
    "from keras import models, layers, activations, optimizers, regularizers\n",
    "from keras.utils import plot_model\n",
    "from keras.models import load_model\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import confusion_matrix\n",
    "import pydot # optional, but required by keras to plot the model "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This tutorial has been tested with:\n",
    "* python 3.6.8\n",
    "* numpy 1.16.3\n",
    "* scipy 1.2.1\n",
    "* keras 2.2.4\n",
    "* tensorflow 1.13.1\n",
    "* scikit-image              0.15.0\n",
    "* scikit-learn              0.20.3\n",
    "* matplotlib                3.0.3\n",
    "* pydot                     1.4.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run -i ImaGene.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's perform the first iteration of training.\n",
    "\n",
    "To do that, the first thing to do is to read the first batch of simulations and store them into an _ImaFile_ object. We assume we have 128 samples (haplotypes) using a 3-epoch demographic model for Europeans."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "myfile = ImaFile(simulations_folder='/home/mfumagal/Data/ImaGene/Binary/Simulations1.Epoch3', nr_samples=128, model_name='Marth-3epoch-CEU')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we populate an _ImaGene_ object by selection the variabile we want to estimate/predict ('selection_coeff_hetero') and how many data points per class we wish to retain. As a quick example, we will use only 2000 data points per class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygene = myfile.read_simulations(parameter_name='selection_coeff_hetero', max_nrepl=2000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can have a look at the data stored in this object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An object of 100 images\n",
      "Rows: min 128, max 128, mean 128.000000, std 0.000000\n",
      "Columns: min 125, max 397, mean 250.790000, std 56.518899\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mygene.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like we have 10000 images in this object. Recall that with the first line we simulated 5 classes and retain 2000 data points for each class. All images have 128 rows as expected, as this represents the number of simulated haplotypes. However, images have different number of columns, ranging from 110 to 422 with an average value of around 257. The number of columns represents the number of polymorphic sites and fixed derived alleles in a _msms_ file. This number may vary from simulated image to another.\n",
    "\n",
    "We can also check the sample allele frequency for the selected allele. recall that we imposed selection to be acting in the middle of the region. Therefore, the targeted allele will be in position '0.5' in the _msms_ file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Allele frequency')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEKCAYAAAAIO8L1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAHaxJREFUeJzt3XuUnVWZ5/Hvj7KAwlsRiQgV6ACdiQ0ECdYISmsjuDqAaEpEBW07Kj1ZPWN76zFKRlcHWhzAtNcexYmKxNbmIoaAlybQCG3PYohWEiAgFmYQQipIgpCATQmh8swf765w6vCeU6eqzv38PmudVefd5z1nP7VTlaf2u9+9tyICMzOzYns1OgAzM2tOThBmZpbLCcLMzHI5QZiZWS4nCDMzy+UEYWZmuZwgzMwslxOEmZnlcoIwM7NcL2h0ANNxwAEHxOzZsxsdhplZS1m3bt2jETFzovNaOkHMnj2bwcHBRodhZtZSJD1YyXm+xGRmZrmcIMzMLJcThJmZ5XKCMDOzXE4QZmaWq6XvYjKz1nfMsht44unRPccv2aeLuy44tYERNbfZ5/34eWUPXPzmmtTlHoSZNUxxcgB44ulRjll2Q4Miam55yaFc+XQ5QZhZwxQnh4nKrb6cIMzMLJcThJmZ5XKCMDOzXE4QZmYtokuaVPl0OUGYmbWIc44/ZFLl0+V5EGZmLeLCgXkAXLH2IUYj6JI45/hD9pRXmxOEmVkLuXBgXs0SQjFfYjIzs1xOEGZmlqtmCULSZZK2Sbq7oGy5pF9JukvStZJ6C15bKmmTpCFJC2oVl5mZVaaWPYjLgeIVt24Cjo6IY4D7gKUAko4EzgaOSu/5mqSuGsZmZmYTqFmCiIifAY8Vld0YEc+mw9uBWen5QuDKiHg6In4DbAJeU6vYzKw51Pu+fpucRo5BfAD4l/S8D3io4LUtqczM2li97+u3yWlIgpD0KeBZ4HtjRTmnRYn3LpY0KGlw+/bttQrRzOrgwoF5nHjEjHFlJx4xo263cVp5dU8QkhYBZwDviYixJLAFKPyTYRawNe/9EbEiIvojon/mzJm1DdbMamr1hmHWb945rmz95p2s3jDcoIisUF0ThKRTgU8Cb42Ipwpeuh44W9I+kg4D5gA/r2dsZlZ/y9cMMbJr/N4PI7tGWb5mqEERWaGazaSWdAVwEnCApC3AMrK7lvYBblI2CHV7RPx1RNwj6Wrgl2SXnj4YEd4xxKzNbd0xMqlyq6+aJYiIOCen+Ftlzv8s8NlaxWNmzefg3h6Gc5LBwb09DYjGinkmtZk1zJIFc+npHj/lqae7iyUL5jYoIivkxfrMrGEG5md3sy9fM8TWHSMc3NvDkgVz95RbYzlBmFlDDczvc0JoUr7EZGZmuZwgzMwslxOEmZnlcoIwM7NcHqQ2q7JPr95Ytz2DzWrJCcKsij69eiPfvX3znuPRiD3HThLWanyJyayKrlj70KTKzZqZE4RZFY1G7ir1JcvNmpkThJmZ5XKCMDOzXE4QZmaWywnCzMxyOUGYmVkuJwgzM8vlBGFmZrmcIMzMLJcThJmZ5XKCMKui/ffrnlS5WTNzgjCromVvOYquvTSurGsvsewtRzUoIrOpq1mCkHSZpG2S7i4omyHpJkm/Tl/3T+WS9BVJmyTdJem4WsVlVmvFv1T+K8xaVS1/di8HTi0qOw+4OSLmADenY4DTgDnpsRi4tIZxmdXM8jVD7No9fmG+XbuD5WuGGhSR2dTVLEFExM+Ax4qKFwIr0/OVwEBB+XciczvQK+mgWsVmVitbd4xMqtysmdW793tgRDwMkL6+PJX3AYUL5m9JZc8jabGkQUmD27dvr2mwZpP10p78wehS5WbNrFkujyqnLHcB/YhYERH9EdE/c+bMGodlNjnK+0kuU27WzOqdIB4Zu3SUvm5L5VuAQwrOmwVsrXNsZtP2+FO7JlVu1szqnSCuBxal54uA6wrK/zLdzXQCsHPsUpRZK+kq0VUoVW7WzF5Qqw+WdAVwEnCApC3AMuBi4GpJ5wKbgXek038CnA5sAp4C3l+ruMxqyVuOWjupWYKIiHNKvHRKzrkBfLBWsZjVS19vD8M5dyz19fY0IBqz6WmWQWqztrBkwVx6urvGlfV0d7FkwdwGRWQ2dTXrQZh1ooH52d3Zy9cMsXXHCAf39rBkwdw95WatxD0IMzPL5R6EWRWt3jDM0lUbGdk1CsDwjhGWrtoI4F6EtRz3IMyqaPmaoT3JYczIrlGvxWQtyQnCrIq8FpO1EycIsyo6uMTtrKXKzZqZE4RZFS1ZMJfuog2DuveSb3O1luQEYVZtxatqeJUNa1FOEGZVtHzNELtGizYMGvWGQdaaJkwQkromOsfMMh6ktnZSSQ9ik6Tlko6seTRmLc6D1NZOKkkQxwD3Ad+UdHva0e0lNY7LrCUtWTCX7q6iQeouD1Jba5owQUTEkxHxjYh4HfAJsmW7H5a0UtIf1zxCs1ZTvLK3V/q2FlXRGISkt0q6Fvgy8HngcOCHZPs4mFmyfM0Qu3YXDVLv9iC1taZK1mL6NXALsDwibisov0bSG2oTlllr8iC1tZNKEsQxEfH7vBci4sNVjsespfXu1527/3Tvft0NiMZseioZpP6qpN6xA0n7S7qshjGZtaxSO4t6x1FrRRXdxRQRO8YOIuJxYH7tQjJrXTtHnt97KFdu1swqSRB7Sdp/7EDSDLyPhFkuz4OwdlJJgvg8cJukz0j6DHAb8LnahmXWmrwntbWTCXsCEfEdSeuAN5ItO3ZmRPyy5pGZtSDvSW3tpNJLRb8CHh87X9KhEbF5qpVK+hjwV2RTiDYC7wcOAq4EZgDrgfdGxDNTrcOsUQbm9zkhWFuoZKLch4BHgJuAHwE/Tl+nRFIf8GGgPyKOBrqAs4FLgC9GxByyZHTuVOswM7Ppq6QH8RFgbkT8rsr19kjaBewHPAycDLw7vb4SOB+4tIp1mpnZJFQySP0QsLNaFUbEMPAPwGayxLATWAfsiIhn02lbgNw+eloscFDS4Pbt26sVlpmZFamkB3E/cKukHwNPjxVGxBemUmG6ZXYhcBiwA/g+cFrOqblTiyJiBbACoL+/39OPzMxqpJIEsTk99k6P6XoT8JuI2A4gaRXwOqBX0gtSL2IWsLUKdZmZ2RRVcpvrBQCSXhgR/1GFOjcDJ0jaDxgBTgEGyRYEPIvsTqZFwHVVqMvMzKaokruYXivpl8C96fhVkr421QojYi1wDdmtrBtTDCuATwJ/K2kT8DLgW1Otw8zMpq+SS0xfAhYA1wNExJ3TXeY7IpaRbTxU6H7gNdP5XDMzq55K7mIiIh4qKhqtQSxmZtZEKulBPCTpdUBI2ptsktu9tQ3LrHWt3jDspTasLVSSIP6abKvRPrL5CTcCH6xlUGatavWGYZau2sjIrqyTPbxjhKWrNgI4SVjLqeQupkeB99QhFrOWt3zN0J7kMGZk1yjL1ww5QVjLmTBBSPo2OZPWIuIDNYnIrIV5T2prJ5VcYipcmG9f4G14EptZroN7exjOSQbeMMha0YR3MUXEDwoe3wPeCRxd+9DMWo83DLJ2MpWtQ+cAh1Y7ELN24A2DrJ1UMgbxJNkYhNLX35LNejazHN4wyNpFJXcxvbgegZiZWXOppAdxXLnXI2J99cIxM7NmUckYxNeA44C7yC4zHQOsBXaRXXI6uWbRmbUgz6S2dlFJgngA+C8RsRFA0tHAxyPifTWMy6wleSa1tZNKFut75VhyAIiIu4FjaxeSWesqN5ParNVU0oO4V9I3ge+SXVL6C7xYn1kuz6S2dlJJD+L9wD3AR4CPAr9MZWZWpNSMac+ktlZUyUzqPwBfB86LiLdFxBdTmZkV8UxqayeVbDn6VuAO4IZ0fKyk62sdmFkrGpjfx0VnzqOvtwcBfb09XHTmPA9QW0uqZAxiGdlWoLcCRMQdkmbXLiSz1uaZ1NYuKhmDeDYidtY8EjMzayqV9CDulvRuoEvSHLItR2+rbVhmZtZolfQgPgQcBTwN/DOwk+xuJjMza2NlexCSuoALImIJ8KlqVSqpF/gm2b4SAXwAGAKuAmaTzd5+Z0Q8Xq06zcxscsr2ICJiFHh1Der9MnBDRLwSeBXZxLvzgJsjYg5wczo2M7MGqWQMYkO6rfX7wH+MFUbEqqlUKOklwBuA96XPeQZ4RtJC4KR02kqyu6a874SZWYNUkiBmAL9j/KqtAUwpQQCHA9uBb0t6FbCObJb2gRHxMEBEPCzp5XlvlrQYWAxw6KHe2M7MrFZKJghJl0TEJ4GfRMT3q1znccCHImKtpC8zictJEbECWAHQ398fVYzLzMwKlBuDOF1SN7C0ynVuAbZExNp0fA1ZwnhE0kEA6eu2KtdrZmaTUC5B3AA8Chwj6YmCx5OSnphqhRHxW+AhSWOL05xCtgDg9cCiVLYIuG6qdZiZ2fSVvMSUbm1dIum6iFhY5Xo/BHxP0t7A/WSrw+4FXC3pXGAz8I4q12lmZpMw4SB1DZIDEXEH0J/z0inVrsvMzKamkpnUZmbWgZwgzMwsV0UJQlJPwaCymZl1gEo2DHoL3jDIzKzjVNKDOJ9sw6AdsGeAeXbtQjIzs2bgDYPMzCyXNwwyM7Nck90w6ArgCbxhkJlZ26tkotxTZJsFVW3DIDMza37lVnP9Idmy3rki4q01icjMzJpCuR7EP9QtCrM2snrDMMvXDLF1xwgH9/awZMFcBub3NToss0krt1jfv409l9QDHBoRQ3WJyqxFrd4wzNJVGxnZNQrA8I4Rlq7aCOAkYS3HE+XMqmj5mqE9yWHMyK5Rlq/x31bWejxRzqyKtu4YmVS5WTPzRDmzKjq4t2dS5WbNrJIEMW6inKR/xBPlzHItWTCXnu6ucWU93V0sWeC1Lq31eKKcWRUNzO/jojPn0dfbg4C+3h4uOnOeB6itJSmi5FSHptff3x+Dg4ONDsPMrKVIWhcRebt6juOJcmZmlssT5WxCnvhl1pkqmihnncsTv8w6V7lLTBvJv8QkICLimOlULKkLGASGI+IMSYcBVwIzgPXAeyPimenUYdNXbuKXE4RZeyt3iemMGtf9EeBe4CXp+BLgixFxpaSvA+cCl9Y4BpuAJ36Zda6St7lGxIN5D2AW8InpVCppFvBm4JvpWMDJwDXplJXAwHTqsOrwxC+zzlXJPIix9Zc+J+kB4ELgV9Os90tkSWZ3On4ZsCMink3HWwBfv2gCnvhl1rnKjUH8J+Bs4Bzgd8BVZPMm3jidCiWdAWyLiHWSThorzjk19xZbSYuBxQCHHnrodEKxCoyNM/guJrPOU3KinKTdwL8D50bEplR2f0QcPq0KpYuA9wLPAvuSjUFcCywAXhERz0p6LXB+RCwo91meKGdmNnmVTpQrd4np7cBvgVskfUPSKeT/pT8pEbE0ImZFxGyyHspPI+I9wC3AWem0RcB1063LzMymrtw8iGuBayW9kGzA+GPAgZIuBa6NiBurHMsngSslXQhsAL5V5c+3KfJEObPONKm1mCTNAN4BvCsiTq5ZVBXyJabaK54oB9kgtRegM2td1bjE9DwR8VhE/O9mSA5WH94hzaxzTSpBWOcZLjEhrlS5mbUPJwgrq0v59yWUKjez9uEEYWWNlhijKlVuZu3DCcLK2n+/7kmVm1n7cIKwskp1FNyBMGt/ThBW1s6RXZMqN7P24QRhZb20J/9SUqlyM2sfThBWVqmblXwTk1n7c4KwsnY8lX8pqVS5mbUPJwgryxsGmXUuJwgra8mCuXR3jb+e1N0lbxhk1gGcIGxCo6NR9tjM2pMThJV1/vX37NkXdszuVG5m7c0JwsraUWK+Q6lyM2sfJTcMMrOp8QZL1i6cIKys/ffr5vGcW1q9FlO+4g2WhneMsHTVRgAnCWs5vsRkZS17y1HsVTQpbi9l5fZ83mDJ2okThE1IRdOmi4/tOVtLbKRUqtysmTlBWFkX/PAeRncX3ea6O7jgh76LKY8nFlo7cYKwsvLGH8qVd7olC+bS0901rqynu8sTC60leZDarIrGBqJ9F5O1g7onCEmHAN8BXkE252pFRHxZ0gzgKmA28ADwzoh4vN7x2Xg93Xsxsqt4qlxWbvkG5vc5IVhbaMRv+bPAf4+IPwFOAD4o6UjgPODmiJgD3JyOrcH2LbpcMlG5mbWPuieIiHg4Itan508C9wJ9wEJgZTptJTBQ79js+bzct1nnauh1AkmzgfnAWuDAiHgYsiQCvLxxkdkY35Vj1rkaliAkvQj4AfDRiHhiEu9bLGlQ0uD27dtrF6AB2V05XUUz5br28nLfZp2gIQlCUjdZcvheRKxKxY9IOii9fhCwLe+9EbEiIvojon/mzJn1CbiDDT74WO48iMEHH2tQRGZWL3VPEMqm4X4LuDcivlDw0vXAovR8EXBdvWOz57ti7UOTKjez9tGIeRAnAu8FNkq6I5X9D+Bi4GpJ5wKbgXc0IDYrMhr5mwOVKjez9lH3BBER/wcotZjPKfWMxczMSvNsJyurVCb3cn1m7c8JwsoqdSHJF5jM2p8ThJmZ5XKCMDOzXE4QVlZfiRnTpcrNrH04QVhZ3t/ArHN5Pwgry/sbmHUu9yBsQoMPPsZvd/6BAH678w9eZsOsQ7gHYWV9evVGvnv75j3HoxF7ji8cmNeosMysDtyDsLK8FpNZ53KCsLK8FpNZ53KCsLK6lL+oRqlyM2sfThBW1jnHHzKpcjNrHx6ktrLGBqKvWPsQoxF0SZxz/CEeoDbrAIoWvpbc398fg4ODjQ7DzKylSFoXEf0TnecehE3o06s3ugdh1oGcIKwsz4Mw61wepLayPA/CrHM5QVhZngdh1rmcIKwsz4Mw61wdNwbhAdfJOef4Q8aNQRSWm1l766gexNiA69jlkbEB10+v3tjgyJrXhQPzOPGIGePKTjxihpOqWQdougQh6VRJQ5I2STqvmp+d95dwuXKD1RuGWb9557iy9Zt3snrDcIMiMrN6aaoEIakL+CpwGnAkcI6kIxsbVWdbvmaIkV2j48pGdo2yfM1QgyIys3ppqgQBvAbYFBH3R8QzwJXAwgbH1NG27hiZVLmZtY9mSxB9QOEN9ltSmTXIwb09kyo3s/bRbAki797JcTfcS1osaVDS4Pbt2+sUVudasmAuPd1d48p6urtYsmBugyIys3pptgSxBSi8f3IWsLXwhIhYERH9EdE/c+bMugbXiQbm93HRmfPo6+1BQF9vDxedOY+B+e7YmbW7ZpsH8QtgjqTDgGHgbODdjQ3JBub3OSGYdaCm6kFExLPA3wBrgHuBqyPinmp9/gMXv3lS5WZmnazZehBExE+An9Tq850MzMwq01Q9CDMzax5OEGZmlssJwszMcjlBmJlZLicIMzPLpWjhncEkbQcenOLbDwAerWI41dKscUHzxua4JsdxTU47xvVHETHhTOOWThDTIWkwIvobHUexZo0Lmjc2xzU5jmtyOjkuX2IyM7NcThBmZparkxPEikYHUEKzxgXNG5vjmhzHNTkdG1fHjkGYmVl5ndyDMDOzMjoyQUg6VdKQpE2SzmtwLA9I2ijpDkmDqWyGpJsk/Tp93b8OcVwmaZukuwvKcuNQ5iup/e6SdFyd4zpf0nBqszsknV7w2tIU15CkBTWM6xBJt0i6V9I9kj6SyhvaZmXiamibSdpX0s8l3ZniuiCVHyZpbWqvqyTtncr3Sceb0uuz6xzX5ZJ+U9Bex6byuv3sp/q6JG2Q9KN0XN/2ioiOegBdwP8DDgf2Bu4EjmxgPA8ABxSVfQ44Lz0/D7ikDnG8ATgOuHuiOIDTgX8h2wHwBGBtneM6H/h4zrlHpn/PfYDD0r9zV43iOgg4Lj1/MXBfqr+hbVYmroa2Wfq+X5SedwNrUztcDZydyr8O/Nf0/L8BX0/PzwauqlF7lYrrcuCsnPPr9rOf6vtb4J+BH6XjurZXJ/YgXgNsioj7I+IZ4EpgYYNjKrYQWJmerwQGal1hRPwMeKzCOBYC34nM7UCvpIPqGFcpC4ErI+LpiPgNsIns37sWcT0cEevT8yfJ9i/po8FtViauUurSZun7/n067E6PAE4Grknlxe011o7XAKdIytuSuFZxlVK3n31Js4A3A99Mx6LO7dWJCaIPeKjgeAvlf4FqLYAbJa2TtDiVHRgRD0P2Cw+8vEGxlYqjGdrwb1IX/7KCS3ANiSt15+eT/fXZNG1WFBc0uM3S5ZI7gG3ATWS9lR2RbRRWXPeeuNLrO4GX1SOuiBhrr8+m9vqipH2K48qJudq+BHwC2J2OX0ad26sTE0ReVm3krVwnRsRxwGnAByW9oYGxVKrRbXgpcARwLPAw8PlUXve4JL0I+AHw0Yh4otypOWU1iy0nroa3WUSMRsSxZHvNvwb4kzJ1NywuSUcDS4FXAv8ZmAF8sp5xSToD2BYR6wqLy9Rdk7g6MUFsAQ4pOJ4FbG1QLETE1vR1G3At2S/OI2Pd1vR1W4PCKxVHQ9swIh5Jv9S7gW/w3CWRusYlqZvsP+HvRcSqVNzwNsuLq1naLMWyA7iV7Bp+r6SxnS0L694TV3r9pVR+qXG6cZ2aLtVFRDwNfJv6t9eJwFslPUB2Gfxksh5FXdurExPEL4A56W6AvckGdK5vRCCSXijpxWPPgT8H7k7xLEqnLQKua0R8ZeK4HvjLdEfHCcDOscsq9VB0zfdtZG02FtfZ6Y6Ow4A5wM9rFIOAbwH3RsQXCl5qaJuViqvRbSZppqTe9LwHeBPZ+MgtwFnptOL2GmvHs4CfRhqBrUNcvypI8iK7zl/YXjX/d4yIpRExKyJmk/0f9dOIeA/1bq9qjba30oPsToT7yK6BfqqBcRxOdgfJncA9Y7GQXTu8Gfh1+jqjDrFcQXbpYRfZXyPnloqDrDv71dR+G4H+Osf1T6neu9IvxkEF538qxTUEnFbDuP6UrAt/F3BHepze6DYrE1dD2ww4BtiQ6r8b+LuC34Gfkw2Ofx/YJ5Xvm443pdcPr3NcP03tdTfwXZ6706luP/sFMZ7Ec3cx1bW9PJPazMxydeIlJjMzq4AThJmZ5XKCMDOzXE4QZmaWywnCzMxyOUFY05H0qbSy5l1pJc3jJzj/cklnlTunxPtmS3p3wXG/pK9MJeZakXRFaoePSXplao8Nko6QdNsE7/17SW+aYr3HqmDFV+tML5j4FLP6kfRa4AyyFUmflnQA2aq7tTAbeDfZaplExCAwWKO6Jk3SK4DXRcQfpePzgOsiYlk65XXl3h8RfzeN6o8F+oGfTOMzrMW5B2HN5iDg0ciWOCAiHo20HImkV0v6t7Sw4Zq8VTRLnSPpjyX9q7J1/9dLOgK4GHh9+qv8Y5JO0nPr7s+QtDr99X67pGNS+flpsbtbJd0v6cN534SyPUfWp/punuAzX5g+8xepdzC2uvCNwMtTfMuAjwJ/JemW9L7fF9T3CWX7itwp6eJUtqdnVaZdbpV0ibI9Ee6T9HplKwz8PfCuVPe7pvyvaa2t1rMA/fBjMg/gRWSzf+8Dvgb8WSrvBm4DZqbjdwGXpeeXky0vUO6ctcDb0vN9gf0omKGayvccA/8ILEvPTwbuSM/PT3XsAxwA/A7oLvoeZpKtrHlYOp4xwWf+T+Av0vPe9L2/kKyHU3IfDOD36etpKab9iuqrpF1uBT6fnp8O/Gt6/j7gfzX658GPxj58icmaSkT8XtKrgdcDbwSuSpdWBoGjgZuy5XHoIluCo9DcvHOUrXfVFxHXpjr+AKDyy+X/KfD2dP5PJb1M0kvTaz+OrIfztKRtwIFky4CMOQH4WWT7KxARj03wmX9OtjDbx9N5+wKHAiMTtVfyJuDbEfFUUX1jctul4PWxhQbXkSUlM8BjENaEImKU7C/bWyVtJFuEbB1wT0S8tsxblXeOpJdMIYxyyyc/XVA2yvN/j0T+UsulPlPA2yNiaNzJlW8bWaq+wtfLtd3Y95P3vVgH8xiENRVJcyXNKSg6FniQbCG5mWkQG0ndko4qenvuOZHth7BF0kAq30fSfsCTZNty5vkZ8J50/klk4yLl9nso9H+BP1O2OiqSZkzwmWuADyn9eS9pfoX1jLkR+ED6ngrrG1NJ2xUr1zbWIZwgrNm8CFgp6ZeS7iLtpxzZ9rBnAZdIupNsnGLcXTwTnPNe4MPpM28DXkG2guezaWD3Y0VxnA/0p/Mv5rmllCcUEduBxcCqFMdVE3zmZ8jGCe6SdHc6rlhE3EC2Quugsp3RPl70+oRtl+MW4EgPUnc2r+ZqZma53IMwM7NcThBmZpbLCcLMzHI5QZiZWS4nCDMzy+UEYWZmuZwgzMwslxOEmZnl+v/+kiDqu0gYEwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "freqs = calculate_allele_frequency(mygene, 0.5)\n",
    "plt.scatter(mygene.targets, freqs, marker='o')\n",
    "plt.xlabel('Selection coefficient')\n",
    "plt.ylabel('Allele frequency')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see, the current allele frequency for the targeted allele (on the y-axis) tends to increase for increasing selection coefficient (on the x-axis).\n",
    "\n",
    "Next, _ImaGene_ provides functionalities to manipulate our object. Specifically we can do the following:\n",
    "* convert ancestral/derived to major/minor allele polarisation\n",
    "* filter out columns based on a minimum allele frequency (e.g. 0.01)\n",
    "* sorting rows and columns by frequency (or distance from the most frequent entry)\n",
    "* resize rows and columns (e.g. to 128x128)\n",
    "\n",
    "For instance, the options above could be achieved with the following lines:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygene.majorminor()\n",
    "mygene.filter_freq(0.01)\n",
    "mygene.sort('rows_freq')\n",
    "mygene.sort('cols_freq')\n",
    "mygene.resize((128, 128))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After the data manipulation is done, we have to convert images to proper _numpy_ float matrices. The following line will do the job. Note that this '.convert' method allows you to normalise the data too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygene.convert(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As an illustration, let's plot one image per class. We can also check the new data dimensions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sel in mygene.classes:\n",
    "    print(sel)\n",
    "    mygene.plot(np.where(mygene.targets == sel)[0][0])\n",
    "mygene.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's suppose that we want to do a simple binary classification. This means that we only want to consider 2 classes. For doing that, first we need to set '.classes' to the desired values and then we need to take a subset of the data corresponding to the desired classes only.\n",
    "\n",
    "We cam achieve these steps with the following lines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygene.classes = np.array([0,300])\n",
    "classes_idx = get_index_classes(mygene.targets, mygene.classes)\n",
    "len(classes_idx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see, we now have 4000 data points, as expected. Finally, let's take the corresponding subset of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygene.subset(classes_idx)\n",
    "mygene.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have an object of 4000 data points (images). \n",
    "\n",
    "The next thing to do is very important. We need to randomly shuffle our retained images before using them for training our network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnd_idx = get_index_random(mygene)\n",
    "mygene.subset(rnd_idx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At the moment our targets represent the 2 possible classes. However, since we are doing a binary classification, we need to vectorise them as required by _keras_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygene.targets = to_binary(mygene.targets)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The object is now ready to be used for the classification!\n",
    "You can save it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygene.save(file='mygene')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want to load an _ImaGene_ object you can use the following function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygene = load_imagene(file='mygene')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that our data is ready, we can build our network.\n",
    "Specifically, we can build a model in _keras_ with convolutional, pooling and dense layers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = models.Sequential([\n",
    "                    layers.Conv2D(filters=32, kernel_size=(3,3), strides=(1,1), activation='relu', kernel_regularizer=regularizers.l1_l2(l1=0.005, l2=0.005), padding='valid', input_shape=mygene.data.shape[1:4]),\n",
    "                    layers.MaxPooling2D(pool_size=(2,2)),\n",
    "                    layers.Conv2D(filters=32, kernel_size=(3,3), strides=(1,1), activation='relu', kernel_regularizer=regularizers.l1_l2(l1=0.005, l2=0.005), padding='valid'),\n",
    "                    layers.MaxPooling2D(pool_size=(2,2)),\n",
    "                    layers.Conv2D(filters=32, kernel_size=(3,3), strides=(1,1), activation='relu', kernel_regularizer=regularizers.l1_l2(l1=0.005, l2=0.005), padding='valid'),\n",
    "                    layers.MaxPooling2D(pool_size=(2,2)),\n",
    "                    layers.Flatten(),\n",
    "                    layers.Dense(units=64, activation='relu'),\n",
    "                    layers.Dense(units=1, activation='sigmoid')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, let's compile our _keras_ model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.compile(optimizer='rmsprop',\n",
    "              loss='binary_crossentropy',\n",
    "              metrics=['accuracy'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at a summary of the model and plot it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.summary()\n",
    "plot_model(model, 'net.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready for doing the training on this first batch of data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "score = model.fit(mygene.data, mygene.targets, batch_size=32, epochs=1, verbose=1, validation_split=0.10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Recall that you can save a _keras_ model with `model.save('net.h5')`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can initialise a network object _ImaNet_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mynet = ImaNet(name='[C32+P]x3+D64')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can keep track of scores (loss and accuracy) across iterations with '.update_scores'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mynet.update_scores(score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OK, now we can repeat the whole procedure described above using all remaning batches of data, leaving the last one for testing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 2\n",
    "while i < 10:\n",
    "\n",
    "    print(i)\n",
    "    \n",
    "    myfile = ImaFile(simulations_folder='/home/mfumagal/Data/ImaGene/Binary/Simulations' + str(i) + '.Epoch3', nr_samples=128, model_name='Marth-3epoch-CEU')\n",
    "    mygene = myfile.read_simulations(parameter_name='selection_coeff_hetero', max_nrepl=2000)\n",
    "\n",
    "    mygene.majorminor()\n",
    "    mygene.filter_freq(0.01)\n",
    "    mygene.sort('rows_freq')\n",
    "    mygene.sort('cols_freq')\n",
    "    mygene.resize((128, 128))\n",
    "    mygene.convert()\n",
    "\n",
    "    mygene.classes = np.array([0,300])\n",
    "    mygene.subset(get_index_classes(mygene.targets, mygene.classes))\n",
    "    mygene.subset(get_index_random(mygene))\n",
    "\n",
    "    mygene.targets = to_binary(mygene.targets)\n",
    "     \n",
    "    score = model.fit(mygene.data, mygene.targets, batch_size=32, epochs=1, verbose=1, validation_split=0.10)\n",
    "    mynet.update_scores(score)\n",
    "   \n",
    "    i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can plot loss and validation accuracy during the training to check, for instance, for overfitting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mynet.plot_train()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We save (and/or load) the final trained model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.save('net.h5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = load_model('net.h5')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we evaluate the training on the testing dataset, i.e. the last batch of simulated data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 10\n",
    "myfile = ImaFile(simulations_folder='/home/mfumagal/Data/ImaGene/Binary/Simulations' + str(i) + '.Epoch3', nr_samples=128, model_name='Marth-3epoch-CEU')\n",
    "mygene_test = myfile.read_simulations(parameter_name='selection_coeff_hetero', max_nrepl=2000)\n",
    "\n",
    "mygene_test.majorminor()\n",
    "mygene_test.filter_freq(0.01)\n",
    "mygene_test.sort('rows_freq')\n",
    "mygene_test.sort('cols_freq')\n",
    "mygene_test.resize((128, 128))\n",
    "mygene_test.convert()\n",
    "\n",
    "mygene_test.classes = np.array([0,300])\n",
    "classes_idx = get_index_classes(mygene_test.targets, mygene_test.classes)\n",
    "mygene_test.subset(classes_idx)\n",
    "rnd_idx = get_index_random(mygene_test)\n",
    "mygene_test.subset(rnd_idx)\n",
    "\n",
    "mygene_test.targets = to_binary(mygene_test.targets)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's report loss and accuracy on the testing set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mynet.test = model.evaluate(mygene_test.data, mygene_test.targets, batch_size=None, verbose=0)\n",
    "print(mynet.test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a binary or multiclass classification, it is convenient to plot the confusion matrix after predicting the responses from the testing data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mynet.predict(mygene_test, model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mynet.plot_cm(mygene_test.classes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the estimation of continuous variabiles, we can produce a scatter plot with `mynet.plot_scatter()`."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
